### PR TITLE
[SWDEV-316198] add split_4g_func in asm igemm bwd solver

### PR DESCRIPTION
**Pull request by shaojiewang**
[asm igemm][NHWC]
Fix solver bug when input/output tensor is over 4GB.
- [x] basic solution for bwd
- [x] corner cases: split-k kernels' 4G
- [x] apply for FWD/WRW
- [x] add test cases

test cases:
```
export MIOPEN_DEBUG_FIND_ONLY_SOLVER="ConvAsmImplicitGemmGTCDynamicFwdXdlopsNHWC;ConvAsmImplicitGemmGTCDynamicBwdXdlopsNHWC;ConvAsmImplicitGemmGTCDynamicWrwXdlopsNHWC"

make -j test_conv2d
./bin/test_conv2d --float --cmode conv --pmode default --group-count 1 --disable-backward-weights --disable-backward-data --disable-verification-cache -v --input 4096 1 512 1024 --weights 1 1 1 1 --batch_size 4096 --input_channels 1 --output_channels 1 --spatial_dim_elements 512 1024 --filter_dims 3 3 --pads_strides_dilations 0 0 1 1 1 1 --trans_output_pads 0 0 --in_layout NHWC --fil_layout NHWC --out_layout NHWC
./bin/test_conv2d --float --cmode conv --pmode default --group-count 1 --disable-backward-weights --disable-forward --disable-verification-cache -v --input 4096 1 512 1024 --weights 1 1 1 1 --batch_size 4096 --input_channels 1 --output_channels 1 --spatial_dim_elements 512 1024 --filter_dims 3 3 --pads_strides_dilations 0 0 1 1 1 1 --trans_output_pads 0 0 --in_layout NHWC --fil_layout NHWC --out_layout NHWC
./bin/test_conv2d --float --cmode conv --pmode default --group-count 1 --disable-forward --disable-backward-data --disable-verification-cache -v --input 4096 1 512 1024 --weights 1 1 1 1 --batch_size 4096 --input_channels 1 --output_channels 1 --spatial_dim_elements 512 1024 --filter_dims 3 3 --pads_strides_dilations 0 0 1 1 1 1 --trans_output_pads 0 0 --in_layout NHWC --fil_layout NHWC --out_layout NHWC

```

### CODE REVIEW

**Comment by junliume** [comment]
@shaojiewang  could you keep updating this PR? The issue is impacting coming releases. Thanks!

**Comment by shaojiewang** [comment]
> 

Yes, I'm still working on it. Basic cases can provide right results, but there exists some corner cases. Furthermore, FWD/WRW also have same problem.
- [x] basic solution for bwd
- [ ] corner cases: split-k kernels' 4G
- [ ] apply for FWD/WRW

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 1]
```diff
@@ -225,6 +225,42 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
...
+    //     return 1;
+    assert(splited_n != 0);
+    while(splited_n >= 1){
+        // printf("n:%d, splited_n:%d\n", n, splited_n);
+        if(n % splited_n == 0 && splited_n * image_size < size_4g)
+            break;
+        splited_n--;
+    }
+    assert(splited_n * image_size < size_4g && n % splited_n == 0);
+    return n / splited_n;
```
Comment (junliume) [inline src/include/miopen/conv/asm_implicit_gemm.hpp:261]: clang tidy is complaining about this line, potential divide by zero. @shaojiewang  anyway to fix it soon?

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 2]
```diff
@@ -225,6 +225,42 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
...
+    size_t image_size = image_size_input >= image_size_output ? image_size_input : image_size_output;
+    size_t splited_n = size_4g / image_size;
+
+    // round up splits, we must match
+    // 1. splited_n * image_size < size_4g
+    // 2. n % splited_n == 0
+    // if(splited_n >= n)
+    //     return 1;
+    assert(splited_n != 0);
+    while(splited_n >= 1){
```
Comment (junliume) [inline src/include/miopen/conv/asm_implicit_gemm.hpp:254]: Would the performance for this loop under control?

Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp:254]: I'll fix this two comments together.

File: src/include/miopen/conv/asm_implicit_gemm.hpp (hunk 1 above)
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp:261]: Yes, I'll fix it very soon.

File: src/include/miopen/conv/asm_implicit_gemm.hpp (hunk 2 above)
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp:254]: @junliume This loop will run at most ```n``` times. But for the most cases, it will run only once. As we need to find the biggest ```splited_n``` under the condition ```n%splited_n==0```, this way should be the fastest.

File: src/include/miopen/conv/asm_implicit_gemm.hpp (hunk 1 above)
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp:262]: Warning fixed.

**Review by carlushuang — DISMISSED** [review]
LGTM, this is to enalbe 4G split in asm igemm kernel.

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 3]
```diff
@@ -225,6 +225,43 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
...
+                                         const int n,
+                                         const int k,
+                                         const int c,
+                                         const int data_byte)
+{
+    size_t image_size_input = static_cast<size_t>(c) * hi * wi * data_byte;
+    size_t image_size_output = static_cast<size_t>(k) * ho * wo * data_byte;
+    size_t size_4g = 0xffffffffUL;
+    if(image_size_input >= size_4g || image_size_output >= size_4g)
+        return 0;
```
Comment (atamazov) [inline src/include/miopen/conv/asm_implicit_gemm.hpp:243]: What is the semantics of return value 0? Please describe it in the code.

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 4]
```diff
@@ -225,6 +225,43 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
...
+
+    // round up splits, we must match
+    // 1. splited_n * image_size < size_4g
+    // 2. n % splited_n == 0
+    // if(splited_n >= n)
+    //     return 1;
+    if(splited_n == 0)
+        return 0;
+    while(splited_n >= 1){
+        if(n % splited_n == 0 && splited_n * image_size < size_4g)
```
Comment (atamazov) [inline src/include/miopen/conv/asm_implicit_gemm.hpp:256]: When `splited_n == 1` this folds into
```
if(n % 1 == 0 && 1 * image_size < size_4g)
```
which is equivalent to
```
if(0 == 0 && true)
```
which means that:
- splited_n can't reach 0 in this loop, and thus lines 260-261 are dead code
- the loop header can be changed to `while(splited_n > 1)`

Comment (atamazov) [reply src/include/miopen/conv/asm_implicit_gemm.hpp:256]: Another problem.

- AFAIU we want to find the biggest value `splited_n` in the range `[1...(size_4g / image_size)]` so that `n % splited_n == 0`. 
- The biggest possible number is `(size_4g / image_size)`, if `n` is divisible by this number without remainder.
- However this algorithm will refuse such a number when `size_4g % image_size == 0`.

Why: in the case described above, `splited_n * image_size == size_4g` and the 2nd part of the following condition is not met:
```
if(n % splited_n == 0 && splited_n * image_size < size_4g)
```
Is `<` correct?

**Review by atamazov — CHANGES_REQUESTED** [review]

**Comment by atamazov** [comment]
Does it require re-tuning of the affected solvers

/cc @JehandadKhan

**Comment by atamazov** [comment]
@junliume Do we want to fix the issue in 5.0 ?

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 5]
```diff
@@ -225,6 +225,43 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
         return std::make_tuple(0, 0, 0);
 }
 
+// this is to support big tensor > 4G. need to decide how many splits needed
+// return the number of splits
```
Comment (atamazov) [inline src/include/miopen/conv/asm_implicit_gemm.hpp:229]:
```suggestion
// This is to support big tensor > 4G. Need to decide how many splits needed.
// Return the number of splits.
```
Ditto other new comments.

**Comment by shaojiewang** [comment]
OCL BE has some problems...
@atamazov Thanks for reviewing, I'll address all of your comments once OCL BE goes well.

**Comment by junliume** [comment]
> OCL BE has some problems... @atamazov Thanks for reviewing, I'll address all of your comments once OCL BE goes well.

CI tests have issues with `test_conv_igemm_dynamic_xdlops_nhwc_` in all three directions as well.

**Comment by shaojiewang** [comment]
> > OCL BE has some problems... @atamazov Thanks for reviewing, I'll address all of your comments once OCL BE goes well.
> 
> CI tests have issues with `test_conv_igemm_dynamic_xdlops_nhwc_` in all three directions as well.

Our gfx90a node seems offline suddenly...

**Comment by JehandadKhan** [comment]
> Does it require re-tuning of the affected solvers?

That is a question for the author. @shaojiewang  Can you please weigh in ?

**Comment by junliume** [comment]
> @junliume Do we want to fix the issue in 5.0 ?

You saw that I've gone back and forth on milestone settings :-P 
If there is a cherry-pick window, I would say yes we need it to be in the next release.

**Comment by atamazov** [comment]
> > @junliume Do we want to fix the issue in 5.0 ?
> 
> You saw that I've gone back and forth on milestone settings :-P If there is a cherry-pick window, I would say yes we need it to be in the next release.

I see. I am asking because I really would like to cherry-pick the fix into 5.0 (because we do not want to realease However, the fix for 5.0 should NOT affect tuning data. Perhaps the ideal fix (or, well, workaround) for 5.0 should only change the applicability of the solvers.

@shaojiewang @carlushuang What is your opinion?

**Comment by shaojiewang** [comment]
> > Does it require re-tuning of the affected solvers?
> 
> That is a question for the author. @shaojiewang Can you please weigh in ?

@JehandadKhan Sorry,  I did not fully understand. How to weigh in?

**Comment by shaojiewang** [comment]
@junliume @carlushuang @atamazov It takes several hours to run 1 single conv ctest case when tensor is over 4GB. So I'm thinking about not to add them in CI or enable GPU reference code by default. What do you think about it?

**Comment by JehandadKhan** [comment]
> > > Does it require re-tuning of the affected solvers?
> > 
> > 
> > That is a question for the author. @shaojiewang Can you please weigh in ?
> 
> @JehandadKhan Sorry, I did not fully understand. How to weigh in?

Does this PR affect the tuning parameters of the kernel or changes their meaning, if so then we would need to retune the kernel.

**Comment by shaojiewang** [comment]
> @shaojiewang having some issues on gfx908:
> 
> > FAILED: /src/ocl/convolutionocl.cpp:1561: Backward Weights Convolution cannot be executed due to incorrect params

I'm checking the problem. But I did not reproduce it locally...

File: src/include/miopen/conv/asm_implicit_gemm.hpp (hunk 4 above)
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp:256]: > splited_n can't reach 0 in this loop, and thus lines 260-261 are dead code

Lines 260-261 are dead code according to max_n's possible value. But without them, clang-tidy keeps throwing divided by zero warning in ```return n/max_n```

> the loop header can be changed to while(splited_n > 1)

agreed.

Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp:256]: > Another problem.
> 
> * AFAIU we want to find the biggest value `splited_n` in the range `[1...(size_4g / image_size)]` so that `n % splited_n == 0`.
> * The biggest possible number is `(size_4g / image_size)`, if `n` is divisible by this number without remainder.
> * However this algorithm will refuse such a number when `size_4g % image_size == 0`.
> 
> Why: in the case described above, `splited_n * image_size == size_4g` and the 2nd part of the following condition is not met:
> 
> ```
> if(n % splited_n == 0 && splited_n * image_size < size_4g)
> ```
> 
> Is `<` correct?

Yes, ```splited_n * image_size < size_4g``` will never meet. Remove this condition.

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 6]
```diff
@@ -225,6 +225,43 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
...
+                                         const int data_byte)
+{
+    size_t image_size_input = static_cast<size_t>(c) * hi * wi * data_byte;
+    size_t image_size_output = static_cast<size_t>(k) * ho * wo * data_byte;
+    size_t size_4g = 0xffffffffUL;
+    if(image_size_input >= size_4g || image_size_output >= size_4g)
+        return 0;
+
+    size_t image_size = image_size_input >= image_size_output ? image_size_input : image_size_output;
+    size_t splited_n = size_4g / image_size;
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp]: Yes, renamed.

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 7]
```diff
@@ -225,6 +225,43 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
...
+                                         const int c,
+                                         const int data_byte)
+{
+    size_t image_size_input = static_cast<size_t>(c) * hi * wi * data_byte;
+    size_t image_size_output = static_cast<size_t>(k) * ho * wo * data_byte;
+    size_t size_4g = 0xffffffffUL;
+    if(image_size_input >= size_4g || image_size_output >= size_4g)
+        return 0;
+
+    size_t image_size = image_size_input >= image_size_output ? image_size_input : image_size_output;
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp]: OK, use ```std::max```.

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 8]
```diff
@@ -225,6 +225,43 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
...
+                                         const int wo,
+                                         const int n,
+                                         const int k,
+                                         const int c,
+                                         const int data_byte)
+{
+    size_t image_size_input = static_cast<size_t>(c) * hi * wi * data_byte;
+    size_t image_size_output = static_cast<size_t>(k) * ho * wo * data_byte;
+    size_t size_4g = 0xffffffffUL;
+    if(image_size_input >= size_4g || image_size_output >= size_4g)
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp]: Done.

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 9]
```diff
@@ -225,6 +225,43 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
...
+    size_t image_size = image_size_input >= image_size_output ? image_size_input : image_size_output;
+    size_t splited_n = size_4g / image_size;
+
+    // round up splits, we must match
+    // 1. splited_n * image_size < size_4g
+    // 2. n % splited_n == 0
+    // if(splited_n >= n)
+    //     return 1;
+    if(splited_n == 0)
+        return 0;
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp]: removed.

File: src/include/miopen/conv/asm_implicit_gemm.hpp (hunk 3 above)
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp:247]: add comments to describe the semantics of value 0.

File: src/include/miopen/conv/asm_implicit_gemm.hpp [hunk 10]
```diff
@@ -225,6 +225,43 @@ HeuristicInitMacroTileNoPadGemmK(size_t gemm_m,
...
+                                         const int ho,
+                                         const int wo,
+                                         const int n,
+                                         const int k,
+                                         const int c,
+                                         const int data_byte)
+{
+    size_t image_size_input = static_cast<size_t>(c) * hi * wi * data_byte;
+    size_t image_size_output = static_cast<size_t>(k) * ho * wo * data_byte;
+    size_t size_4g = 0xffffffffUL;
```
[the remark this answers is not in the document]
Comment (shaojiewang) [reply src/include/miopen/conv/asm_implicit_gemm.hpp]: renamed and made it constexpr.

**Comment by junliume** [comment]
@shaojiewang  could you check this case:

> /home/jenkins/workspace/MLLibs_MIOpen_fix_swdev_316198/build/bin/test_conv2d --half --cmode conv --pmode default --group-count 1 --disable-forward --disable-backward-data --input 4096 1 512 1024 --weights 1 1 1 1 --batch_size 4096 --input_channels 1 --output_channels 1 --spatial_dim_elements 512 1024 --filter_dims 3 3 --pads_strides_dilations 0 0 1 1 1 1 --trans_output_pads 0 0 --in_layout NHWC --fil_layout NHWC --out_layout NHWC 

[2022-01-18T19:51:24.894Z] FAILED: -nan

**Review by atamazov — DISMISSED** [review]

**Review by junliume — APPROVED** [review]
